_posts/2024-04-12-vector-search.md

```diff
@@ -0,0 +1,104 @@
+
+
+While the `HNSW` index supports deletes and insertions to the base table after the index has been created, deletes are not immediately reflected internally in the index structure, but are instead marked as deleted, and will only be removed after a certain number of updates have been made. This is a trade-off to maintain the performance of updates, but will lead to the _quality_ of the index deteriorating over time, impacting both speed and accuracy of retrieval. To remedy this you can use the pragma `pragma_hnsw_compact_index('<index name>')` to trigger a recompaction, which will be faster than dropping and recreating the index. 
+
+Additionally, it is generally recommended to create the index __after__ loading a table with data as the initial index creation step can utilize some degree of parallelism to speed up index creation on larger tables.
+
+
+Like the `ART` the `HNSW` index must be able to fit into RAM in its entirety, but will also be persisted to disk as part of the DuckDB database file (if running DuckDB in disk-backed mode) during checkpointing. After restarting or loading a database file with a persisted `HNSW` index, the index will only be lazily loaded back into memory whenever the table the index is created on is first accessed.
+
+## Future Work
+
+The main limitations of the `HNSW` index so far is that it only supports the `FLOAT` type for the fixed-size list elements and only distance metrics corresponding to the three built in distance functions, `array_distance`, `array_inner_product` and `array_cosine_distance`. Additionally the memory allocated by the `HNSW` at runtime is allocated "outside" of the DuckDB memory management system, meaning that it wont respect DuckDB's  `memory_limit` configuration parameter.
+
+However, we still think that the `HNSW` and `vss` extension is a useful addition to the DuckDB extension ecosystem and we're excited to see what the community will do with it. We're also interested in eventually addressing some of the limitations mentioned above and make it even easier to create custom indexes in DuckDB in the future.
+
+## Conclusion
+
+The `vss` extension for DuckDB is a new extension that adds support for creating HNSW indexes on fixed-size list columns in DuckDB, accelerating vector similarity search queries. The extension can currently be installed on the nightly builds of DuckDB by running `INSTALL vss` and will also be available in the upcoming release of DuckDB v0.10.2. The `vss` extension treads new ground for DuckDB extensions by providing a custom index type and we're excited to refine and expand on this functionality going forward. 
+
```

Comment:
What other outcomes would you like to drive with your post? Are you looking for contributors or maybe other companies to sponsor the work? I would mention those here! Maybe something like (and please feel free to change/ignore):

This work was made possible by the sponsorship of a DuckDB Labs customer! If you are interested in specific capabilities, please [reach out to DuckDB Labs](https://duckdblabs.com/contact/). Alternatively, we welcome contributors! Please reach out to the DuckDB Labs team over on [Discord](https://discord.duckdb.org/) to discuss the roadmap.